go/vt/vtgate/engine/recurse_cte.go

```diff
@@ -0,0 +1,145 @@
+
+func (r *RecurseCTE) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	res, err := vcursor.ExecutePrimitive(ctx, r.Seed, bindVars, wantfields)
+	if err != nil {
+		return nil, err
+	}
+
+	// recurseRows contains the rows used in the next recursion
+	recurseRows := res.Rows
+	joinVars := make(map[string]*querypb.BindVariable)
+	for len(recurseRows) > 0 {
+		// copy over the results from the previous recursion
+		theseRows := recurseRows
+		recurseRows = nil
+		for _, row := range theseRows {
+			for k, col := range r.Vars {
+				joinVars[k] = sqltypes.ValueBindVariable(row[col])
+			}
+			rresult, err := vcursor.ExecutePrimitive(ctx, r.Term, combineVars(bindVars, joinVars), false)
+			if err != nil {
+				return nil, err
+			}
+			recurseRows = append(recurseRows, rresult.Rows...)
+			res.Rows = append(res.Rows, rresult.Rows...)
+		}
+	}
+	return res, nil
+}
+
+func (r *RecurseCTE) TryStreamExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	if vcursor.Session().InTransaction() {
+		res, err := r.TryExecute(ctx, vcursor, bindVars, wantfields)
+		if err != nil {
+			return err
+		}
+		return callback(res)
+	}
+	return vcursor.StreamExecutePrimitive(ctx, r.Seed, bindVars, wantfields, func(result *sqltypes.Result) error {
+		err := callback(result)
+		if err != nil {
+			return err
+		}
+		return r.recurse(ctx, vcursor, bindVars, result, callback)
+	})
+}
+
+func (r *RecurseCTE) recurse(ctx context.Context, vcursor VCursor, bindvars map[string]*querypb.BindVariable, result *sqltypes.Result, callback func(*sqltypes.Result) error) error {
+	if len(result.Rows) == 0 {
+		return nil
+	}
+	joinVars := make(map[string]*querypb.BindVariable)
+	for _, row := range result.Rows {
+		for k, col := range r.Vars {
+			joinVars[k] = sqltypes.ValueBindVariable(row[col])
+		}
+
+		err := vcursor.StreamExecutePrimitive(ctx, r.Term, combineVars(bindvars, joinVars), false, func(result *sqltypes.Result) error {
+			err := callback(result)
+			if err != nil {
+				return err
+			}
+			return r.recurse(ctx, vcursor, bindvars, result, callback)
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *RecurseCTE) RouteType() string {
+	return "RecurseCTE"
+}
+
+func (r *RecurseCTE) GetKeyspaceName() string {
+	if r.Seed.GetKeyspaceName() == r.Term.GetKeyspaceName() {
+		return r.Seed.GetKeyspaceName()
+	}
+	return r.Seed.GetKeyspaceName() + "_" + r.Term.GetKeyspaceName()
+}
+
+func (r *RecurseCTE) GetTableName() string {
+	return r.Seed.GetTableName()
+}
+
+func (r *RecurseCTE) GetFields(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return r.Seed.GetFields(ctx, vcursor, bindVars)
+}
+
+func (r *RecurseCTE) NeedsTransaction() bool {
+	return r.Seed.NeedsTransaction() || r.Term.NeedsTransaction()
+}
```

Comment:
AFAIK, this engine primitive only deals with select statements. This should be `noTxNeeded`

Reply:
`WITH` is valid with `SELECT`, `UPDATE` and `DELETE`. might as well leave it open for both and just check what the inputs do, right?

Reply:
Update and Delete are not part of the recursive CTE inputs.
The output of the CTE will be used to perform the Update and Delete,
so CTE output becomes the input for Update/Delete operator/engine.